packages/plexus/src/LayoutManager/Coordinator.js

```diff
@@ -0,0 +1,329 @@
+    if (ok) {
+      this.idleWorkers.push(worker);
+    } else {
+      killWorker(worker);
+    }
+  }
+
+  _getPositions() {
+    const { id, cleaned, status } = this.currentLayout || {};
+    if (id == null) {
+      throw new Error('_getPositions called without a current layout');
+    }
+    const worker = this.idleWorkers.pop() || this._initWorker();
+    this.busyWorkers.push(worker);
+    status.phase = 'positions';
+    status.positionsWorker = worker.id;
+    worker.postMessage({
+      edges: cleaned.edges,
+      meta: {
+        layoutId: id,
+        workerId: worker.id,
+        phase: 'positions',
+      },
+      vertices: cleaned.vertices,
+    });
+  }
+
+  _getEdges() {
+    const { id, cleaned, result, status } = this.currentLayout || {};
+    if (id == null) {
+      throw new Error('_getEdges called without a current layout');
+    }
+    if (!result.vertices) {
+      throw new Error('_getEdges called without known positions');
+    }
+    const worker = this.idleWorkers.pop() || this._initWorker();
+    this.busyWorkers.push(worker);
+    status.phase = 'edges';
+    // TODO(joe): split edge calculation
+    status.edgesWorkers = new Set([worker.id]);
+    worker.postMessage({
+      edges: cleaned.edges,
+      meta: {
+        layoutId: id,
+        workerId: worker.id,
+        phase: 'edges',
+      },
+      vertices: result.vertices,
+    });
+  }
+
+  _handleVizWorkerError = (event: ErrorEvent) => {
+    const target = (event: any).target;
+    const worker = (target: LayoutWorker);
+    let list;
+    let i = this.busyWorkers.indexOf(worker);
+    if (i >= 0) {
+      list = this.busyWorkers;
+    } else {
+      i = this.idleWorkers.indexOf(worker);
+      if (i >= 0) {
+        list = this.idleWorkers;
+      }
+    }
+    if (!list) {
+      console.error('Viz worker onerror from unknown viz worker');
+      console.error(event);
+      return;
+    }
+    list.splice(i, 1);
+    killWorker(worker);
+    console.error('Viz worker onerror');
+    console.error(event);
+  };
+
+  _handleVizWorkerMessageError = (event: MessageEvent) => {
+    // TODO(joe): something more useful
+    const msg = {
+      event,
+      currentLayout: this.currentLayout,
+      errorType: '_handleVizWorkerMessageError',
+    };
+    console.error(msg);
+  };
+
+  _handleVizWorkerMessage = (event: MessageEvent) => {
+    const worker: LayoutWorker = ((event.target: any): LayoutWorker);
+    const workerMessage = ((event.data: any): WorkerMessage);
+    const { type, meta } = workerMessage;
+    this._makeWorkerIdle(worker);
+    const isStale = !this.currentLayout || meta.layoutId !== this.currentLayout.id;
+    if (type === 'error') {
+      console.error(`${isStale ? '(stale) ' : ' '}Error from viz.worker`, event);
+      return;
+    }
+    if (isStale) {
+      return;
+    }
+    if (type === 'layout-error') {
+      // TODO(joe): implement retries with more spacing
+      console.error('layout-error', event);
```

Comment:
Is these TODOs to be fixed before merge?

Reply:
No, they're enhancements